### PR TITLE
Testcase + fix for today's stuck nodes

### PR DESCRIPTION
**Pull request by pmconrad**

### CODE REVIEW

**Comment by oxarbitrage** [comment]
compiling now and will try a full replay to see if it pass.

**Comment by abitmore** [comment]
I think there are some issues, not about this PR, but about the original implementation:
* in `process_bids()`:
```
    share_type to_cover = bdd.current_supply;
    share_type remaining_fund = bad.settlement_fund;
```
Need a check for zero `current_supply` and/or zero `settlement_fund`, otherwise the following `execute_bid()` will fail

* in `asset_publish_feeds_evaluator::do_apply(...)`:
```
         if( !bad.current_feed.settlement_price.is_null()
             && ~price::call_price(asset(mia_dyn.current_supply, o.asset_id),
                                   asset(bad.settlement_fund, bad.options.short_backing_asset),
                                   bad.current_feed.maintenance_collateral_ratio ) < bad.current_feed.settlement_price )
```
Need a check for zero `current_supply` and/or zero `settlement_fund`

**Comment by pmconrad** [comment]
Hm. Will add more tests.

**Comment by pmconrad** [comment]
swan_tests/revive_empty_with_bid already covers the first case, i. e. an asset with empty supply and settlement_fund. It is incomplete insofar as the bid has an insufficient price and is therefore ignored. If the price is sufficient, execute_bid is called (and succeeds), which results in an open call_order belonging to the bidder with 0 debt. This is not nice, but can be fixed easily.

asset_publish_feed_evaluator::do_apply is covered by swan_tests/revive_empty_recovered. The case with 0 collateral is handled correctly in database::revive_bitasset: https://github.com/bitshares/bitshares-core/blob/master/libraries/chain/db_market.cpp#L104 .

**Comment by pmconrad** [comment]
Testnet release is already out, it's more urgent because it's been stuck all day.

**Comment by abitmore** [comment]
I thought a zero denominator or zero amount in a `price` or `price::call_price` would lead to errors. Good if it's not the case.

**Comment by abitmore** [comment]
@pmconrad are you sure `wait_for_hf_core_216()` doesn't pass a maintenance interval? Otherwise `if( bad.has_settlement() )` in `asset_publish_feeds_evaluator::do_apply(...)` will return false, so the following code block won't be covered in the test.

**Comment by abitmore** [comment]
The `process_bids()` call in `database::perform_chain_maintenance(...)` should be guarded with a hard fork time check, or need a check inside `process_bids()`. Otherwise, before the hard fork time, if a bitAsset has been fully settled, it will be revived unexpectedly.

**Comment by pmconrad** [comment]
You were right again. :-(

File: tests/tests/swan_tests.cpp [hunk 1]
```diff
@@ -490,4 +491,31 @@ BOOST_AUTO_TEST_CASE( revive_empty_with_bid )
    }
 }
 
+/** Creates a black swan, settles all debts - asset should be revived in next maintenance
+ */
```
Comment (abitmore) [inline tests/tests/swan_tests.cpp]: This comment doesn't describe the test accurately.

**Comment by abitmore** [comment]
Looks fine now.

**Comment by oxarbitrage** [comment]
glad you guys were able to create a fix, i was not able to be here yesterday to help with it. well done.

**Pull request merged**